sql/src/main/java/io/crate/analyze/MultiSourceSelect.java

```diff
@@ -23,46 +23,95 @@
 
 import io.crate.analyze.relations.*;
 import io.crate.analyze.symbol.Field;
+import io.crate.analyze.symbol.FieldReplacer;
 import io.crate.analyze.symbol.Symbol;
+import io.crate.metadata.Functions;
 import io.crate.metadata.Path;
+import io.crate.metadata.TransactionContext;
 import io.crate.metadata.table.Operation;
 import io.crate.sql.tree.QualifiedName;
 
 import javax.annotation.Nonnull;
 import java.util.*;
-import java.util.stream.Collectors;
+import java.util.function.Function;
 
 public class MultiSourceSelect implements QueriedRelation {
 
-    private final Map<QualifiedName, RelationSource> sources;
-    private final QuerySpec querySpec;
+    private final Map<QualifiedName, AnalyzedRelation> sources;
     private final Fields fields;
     private final List<JoinPair> joinPairs;
     private final Set<Symbol> requiredForQuery;
     private final Set<Field> canBeFetched;
     private final Optional<RemainingOrderBy> remainingOrderBy;
     private QualifiedName qualifiedName;
-
-    public static MultiSourceSelect createWithPushDown(MultiSourceSelect mss, QuerySpec querySpec) {
+    private QuerySpec querySpec;
+
+    /**
+     * Create a new MultiSourceSelect which has it's sources transformed into subQueries which contain
+     * any expressions that only affect a single source.
+     * Example:
+     * <pre>
+     *     select t1.x, t2.y from t1, t2 where t1.x = 10
+     *
+     *     becomes
+     *
+     *     select t1.x, t2.y from
+     *          (select x from t1 where x = 10) t1,
+     *          (select y from t2)
+     * </pre>
+     */
+    public static MultiSourceSelect createWithPushDown(Functions functions,
+                                                       TransactionContext transactionContext,
+                                                       MultiSourceSelect mss,
+                                                       QuerySpec querySpec) {
         RelationSplitter splitter = new RelationSplitter(
             querySpec,
-            mss.sources.values().stream().map(RelationSource::relation).collect(Collectors.toList()),
+            mss.sources.values(),
             mss.joinPairs
         );
         splitter.process();
 
-        for (RelationSource source : mss.sources.values()) {
-            QuerySpec spec = splitter.getSpec(source.relation());
-            source.querySpec(spec);
+        Function<Field, Field> convertFieldToPointToNewRelations = Function.identity();
+        Function<? super Symbol, ? extends Symbol> convertFieldInSymbolsToNewRelations = Function.identity();
```

Comment:
do you think that's necessary? I think the names are already long enough as it is and thought it'd be okay to omit the "pointTo"